### ISSUE TITLE
Possible incorrect kubelet configuration?

### ISSUE THREAD

**Issue by fourpastmidnight** [kubernetes-sigs/sig-windows-tools#114]
Using `PrepareNode.ps1` as a template for my own endeavors to allow Windows Server 2019 2004 (10.0.19041.508) to be configured as a Kubernetes worker node, I found that the configuration in this script for running Kubelet resulted in an error.

Now, I originally based some of my scripts on what can be found in `KubeCluster.ps1` and `KubeClusterHelper.psm1`. And in those scripts, the configuration used for Kubelet is _slightly different_. And the one major difference was that they do not configure Kubelet with the `--config=/var/lib/kubelet/config.yaml` configuration parameter. So on a hunch, I omitted that parameter.

And when I did so, and joined the cluster, _EVERYTHING_ worked. The Windows Server 2019 2004 (10.0.19041.508) server appears as `Ready` on the control plane node:

```
my-user@k8s-master:~$ kubectl get nodes
NAME              STATUS   ROLES    AGE    VERSION
k8s-master        Ready    master   9d     v1.19.3
k8s-worker-w001   Ready    <none>   113s   v1.19.3
```

With this change, I have now SUCCESSFULLY set up a Windows Server 2019 2004 server as a functioning Kubernetes worker node! 

For clarity, here's the function I'm using to install Kubelet as a service in Windows:

```powershell
function Install-Kubelet {
    [CmdletBinding()]
    [OutputType([System.ServiceProcess.ServiceController])]
    Param (
        [string] $KubeConfig = $env:KUBECONFIG,

        #[string] $KubeDnsServiceIpAddress = '10.96.0.10', # <-- No longer needed
        
        [string[]] $FeatureGates
    )

    Process {
        if (!(Test-Path $Script:KubernetesClusterNodeLogPath)) {
            $null = New-Item -ItemType Directory -Path $Script:KubernetesClusterNodeLogPath
        }

        $KubeletSvc = Get-Service 'kubelet' -ErrorAction SilentlyContinue
        if (!$KubeletSvc) {
            $KubeletArgs = @(
                #(Get-Command kubelet.exe -ErrorAction Stop).Source    # <-- Used by KubeClusterHelper.psm1. That didn't use nssm
                #'--windows-service'   # <--  Also used by KubeClusterHelper.psm1 which doesn't use nssm, but which is convenient for also creating the docker 'host' network on service restarts (esp. after a reboot)

                "--cert-dir=`"$(Join-Path $env:SystemDrive var\lib\kubelet\pki)`""
                
                #'--config=/var/lib/kubelet/config.yaml'  # <--- Used by PrepareNode.ps1, but caused a problem...
                #'--cluster-domain=cluster.local'   # <-- Used by KubeClusterHelper.psm1, but this with the one below causes Kube-Proxy DS to not start (I don't know which was the culprit, but neither are needed).
                #"--cluster-dns=`"$KubeDnsServiceIpAddress`""   # <-- Also used by KubeClusterHelper.psm1; see above.
                
                '--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf'
                "--kubeconfig=$KubeConfig"
                "--hostname-override=$($env:COMPUTERNAME.ToLower())"
                '--pod-infra-container-image=kubeletwin/pause'
                '--enable-debugging-handlers'
                '--cgroups-per-qos=false'
                '--enforce-node-allocatable=""'
                '--network-plugin=cni'
                '--resolv-conf=""'
                "--log-dir=`"$Script:KubernetesClusterNodeLogPath`""
                '--logtostderr=false'
                '--image-pull-progress-deadline=20m'
                '--v=6'
            )

            if ($FeatureGates) {
                $KubeletArgs += "--feature-gates=$($FeatureGates -join ',')"
            }

            $StartKubeletPath = (Join-Path $Script:KubernetesClusterNodeInstallationPath StartKubelet.ps1)

            @"
`$KubeletArgs = (Get-Content -Path '/var/lib/kubelet/kubeadm-flags.env' -Raw).Trim('KUBELET_KUBEADM_ARGS=') -replace '^"(.*)"`$','`$1'

if (!(docker network ls -f name=host -q)) {
    docker network create -d nat host
}

`$KubeletCmd = '$((Get-Command 'kubelet.exe' -ErrorAction Stop).Source) `$KubeletArgs $KubeletArgs'

Invoke-Expression `$KubeletCmd
"@ | Set-Content $StartKubeletPath


            nssm.exe install kubelet (Get-Command powershell.exe).Source -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Bypass -File $StartKubeletPath
            nssm.exe set kubelet DisplayName Kubelet
            nssm.exe set kubelet AppDirectory $Script:KubernetesClusterNodeInstallationPath
            nssm.exe set kubelet DependOnService docker

            <#

            New-Service -Name 'kubelet' -StartupType Automatic -DependsOn 'docker' `
                -BinaryPathName "$KubeletArgs"

            #>

            if (!(Get-NetFirewallRule -Name KubeletAllow10250 -ErrorAction SilentlyContinue)) {
                $null = New-NetFirewallRule -Name KubeletAllow10250 -Description "Kubelet Allow 10250" -Action Allow -LocalPort 10250 -Protocol TCP -Enabled True -DisplayName "Kubelet Allow 10250 (TCP)" -ErrorAction Stop
            }

            $KubeletSvc = Get-Service kubelet -ErrorAction Stop
        }

        $KubeletSvc
    }
}
```

**Comment by jsturtevant** [comment]
What was the error?  Did you have a /var/lib/kubelet/config.yaml file present? I am not sure the behavior if that flag is set and you don't have that file.

**Comment by fourpastmidnight** [comment]
So neither `KubeClusterHelper.psm1`, `KubeCluster.ps1`, nor `PerpareNode.ps1` copied  a file `/var/lib/kubelet/config.yaml`, so no, it wasn't present. I do see that `KubeClusterHelper.psm1`, when setting up Kubelet, would pass the parameter `--config=/var/lib/kubelet/config.yml`. Now _that_ did produce an error, because `config.yml` couldn't be found. When I changed it to `config.yaml`, Kubelet started, but I got various errors due to not finding the certificate directory `/etc/kuberenetes/var/lib/kubelet/pki` (or something like that--the point is, even though the `--cert-dir` parameter was _also_ passed, it seemed to be ignored). When I removed the `--config` parameter completely from the arguments passed to kubelet, almost immediately the `kubelet-flannel-windows-ds-*` pod became ready, and shortly thereafter the `kube-proxy-windows-ds-*` pod.

I guess the secondary point is, yeah, the specified file didn't exist, and none of these scripts actually created or somehow retrieved that file. But omitting the parameter results in a working cluster.

I'm also of the opinion that, if some configuration item is not needed, then why pass it? I'm also of the opinion of giving users a choice. So if someone has a configuration file they want to use, then the scripts should accept a configuration file which would result in passing the parameter to kubelet. Again, the key here is KISS&mdash;if the parameter is not needed, don't overcomplicate the arguments passed to kubelet. K8s is hard enough as it is!!
